compiler/rustc_lint/src/empty_iterator.rs

```diff
@@ -0,0 +1,76 @@
+use crate::{EarlyContext, EarlyLintPass, LintContext};
+
+use rustc_ast::{Expr, ExprKind, Lit, LitKind};
+use rustc_errors::fluent;
+
+declare_lint! {
+    /// The `empty_for_iterator` lint checks for empty `for` iterator.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// for i in 10..0 { /* ... */ }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to have a statement that has no effect.
+    pub EMPTY_FOR_ITERATOR,
+    Warn,
+    "empty for iterator"
+}
+
+declare_lint_pass!(EmptyForIterator => [EMPTY_FOR_ITERATOR]);
+
+impl EarlyLintPass for EmptyForIterator {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        match &expr.kind {
+            ExprKind::ForLoop(_, for_expr, _, _) => {
+                let range_span = expr.span.with_hi(for_expr.span.hi());
+                match for_expr.clone().into_inner().kind {
+                    ExprKind::Range(Some(start), Some(end), _) => {
+                        match (start.clone().into_inner().kind, end.clone().into_inner().kind) {
+                            (
+                                ExprKind::Lit(Lit { kind: LitKind::Int(s, _), .. }),
+                                ExprKind::Lit(Lit { kind: LitKind::Int(e, _), .. }),
+                            ) => {
+                                if s >= e {
+                                    cx.struct_span_lint(EMPTY_FOR_ITERATOR, range_span, |lint| {
+                                        lint.build(fluent::lint::empty_for_iterator).emit();
+                                    });
+                                }
+                            }
+                            _ => (),
+                        }
+                    }
+                    ExprKind::MethodCall(_, exp, _, _) => match &exp.peel_parens().kind {
+                        ExprKind::Range(Some(start), Some(end), _) => {
+                            match (start.clone().into_inner().kind, end.clone().into_inner().kind) {
+                                (
+                                    ExprKind::Lit(Lit { kind: LitKind::Int(s, _), .. }),
+                                    ExprKind::Lit(Lit { kind: LitKind::Int(e, _), .. }),
+                                ) => {
+                                    if s >= e {
```

Comment:
```suggestion
                                    if s > e {
```
See https://github.com/rust-lang/rust-clippy/issues/5689#issuecomment-640012737